### ISSUE TITLE
Failed to receive UDP traffic after pod restart

### ISSUE THREAD

**Issue by kammous** [projectcalico/calico#4124]
Incoming UDP traffic is interrupted for ever after pod restart due to a Calico policy that presents NodePort access.

## Expected Behavior
Incoming UDP traffic must resume after pod restart. Otherwise said, Calico policy must behaves the same with old and new pod.

## Current Behavior
Initially, UDP traffic sent to k8s cluster is received by UDP listener pod. 
After failover or pod deletion, new pod doesn't receive any UDP packets.
Further diving shows that a stale conntrack entry prevents UDP packets from being routed to the new pod. Once deleted, UDP traffic is received by the new pod.

## Possible Solution
- Sender uses a different source port
- Manually delete a stale conntrack entry after failover.
- Remove Calico policy and Host Endpoints.

## Steps to Reproduce (for bugs)
1. Create a deployment for a UDP app linstening to incoming UDP packets. We can use `mendhak/udp-listener` container:
```
kubectl create deploy udp-listener --image=mendhak/udp-listener
```
2. Expose our deployment externally using a LoadBalancer Service type:
```
kubectl expose deployment udp-listener --type=LoadBalancer --protocol=UDP --port=5005 --target-port=5005

export VIP=$(kubectl get svc udp-listener -o=jsonpath='{$.status.loadBalancer.ingress[0].ip}')
```
3. On an external node, periodically send UDP packets to our udp-listener
```shell
while true; do echo -n $(date) | nc -n4u -w1 -p 65300 $VIP 5005; sleep 1s; done;
```
4. Check UDP packets are received and logged by udm-listener:
```
kubectl logs -lapp=udp-listener -f
```
5. Apply Calico rules to deny access to NodePort using below GlobalNetworkPolicy and HostEndpoint:
```yaml
cat <<EOF | calicoctl create -f -
apiVersion: projectcalico.org/v3
kind: HostEndpoint
metadata:
  labels:
    interface: customer
  name: k8s-worker
spec:
  interfaceName: eth0
  node: k8s-worker1
---
apiVersion: projectcalico.org/v3
kind: HostEndpoint
metadata:
  labels:
    interface: customer
  name: k8s-worker2
spec:
  interfaceName: eth0
  node: k8s-worker2
---
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
metadata:
  name: deny-nodeports
spec:
  applyOnForward: true
  ingress:
  - action: Deny
    destination:
      ports:
      - 30000:32767
    protocol: TCP
    source: {}
  - action: Deny
    destination:
      ports:
      - 30000:32767
    protocol: UDP
    source: {}
  order: 1100
  preDNAT: true
  selector: interface == 'customer'
  types:
  - Ingress
EOF
```
6. Delete udp-listener pod
```
kubectl delete $(kubectl get po -l app=udp-listener -o name)
```
7. A new pod is launched but it doesn't receive any UDP packet

**Notes and workaround**
- If we remove step 5, we get the expected behavior, meaning UDP packets are received at step 7.
- The traffic is blocked by a stale conntrack entry. The problem can be fixed once entry removed,
Delete stale conntrack udp entry with dport=5005:
```
# conntrack -D -p udp --orig-port-dst 5005
udp      17 29 src=192.168.81.130 dst=172.17.1.143 sport=65301 dport=5005 [UNREPLIED] src=172.17.1.143 dst=192.168.81.130 sport=5005 dport=65301 mark=0 secctx=system_u:object_r:unlabeled_t:s0 use=3
conntrack v1.4.4 (conntrack-tools): 1 flow entries have been deleted.
```
Observed conntrack events (to be laucnhed before previous delete command): 
```
# conntrack -E -p udp --dport 5005
[DESTROY] udp      17 src=192.168.81.130 dst=172.17.1.143 sport=65301 dport=5005 [UNREPLIED] src=172.17.1.143 dst=192.168.81.130 sport=5005 dport=65301
    [NEW] udp      17 30 src=192.168.81.130 dst=172.17.1.143 sport=65301 dport=5005 [UNREPLIED] src=100.64.145.67 dst=172.17.16.2 sport=5005 dport=50560
```
192.168.81.130 : UDP packets sender
172.17.1.143   : VIP allocated by MatelLB to udp-listener LoadBalancer service
100.64.145.67  : New pod IP
- The behavior is similar to a fixed kube-proxy issue kubernetes/kubernetes#59368

## Context
On-prem k8s environment where NodePort access is blocked using Calico Global Network Policy and Host Endpoints.
`LoadBalancer` service type is handled by MetalLB (layer2)

## Your Environment
* Calico version: v3.13.2
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes v1.17.9
* Operating System and version: RHEL 7.9
* kube-proxy mode: iptables
* Install MetalLB v0.9.3 (layer2 mode) to handle LoadBalancer service type on-prem.

**Comment by fasaxc** [comment]
I must be missing something but you've created a policy `deny-nodeports` that blocks traffic to nodeports and you say traffic to node ports is blocked.  Wasn't that what you wanted?

**Comment by kammous** [comment]
> I must be missing something but you've created a policy `deny-nodeports` that blocks traffic to nodeports and you say traffic to node ports is blocked. Wasn't that what you wanted?

thanks for your feedback. 'deny-nodeports' policy aims to prevent clients from getting access to workloads using NODE_IP:NODE_PORT. However those workloads remain accessible using Loadbalancer VIP as you can see at steps 3 and 4. Shouldn't we expect the same behavior after failover or pod restart?

**Comment by fasaxc** [comment]
Calico's firewall is connection based, policy is only consulted at the start of a new stream.  UDP "connections" are decided by Linux's conntrack system;  by default the timer is something like ~10s before the connection is reset.  If you send a packet every 1s then it'll keep the connection alive.

Similarly, kube-proxy's NAT for node ports is conntrack based; if you keep sending packets on the connection then it may never get rebalanced.

**Comment by lmm** [comment]
@kammous I'm going to close this issue for now since it's been a while but if you're still having issues please re-open or create another issue. Thanks!

**Comment by workhorse2020** [comment]
@lmm I am seeing this issue when I deployed Nginx Ingress controller with Mettallb hostnetwork. This is a major issue if there is only a single src / dst ip and src/dst port traffic with a daemonset. 
-> Same node chosen on pod restart
-> Stale entry not allowing traffic to flow through
-> Post timeout or force delete of conntrack entry its working fine..

**Comment by sunfuze** [comment]
@lmm We have met the same issue.

We deployed jaeger agent in every k8s nodes by daemon set mode, and export 6831(udp) host port to receive traces.

After jaeger agent restarting, jaeger agent can't receive traces anymore. If we restart the producer pod, jaeger agent can receive traces.

Tcpdump in producer pod show data is transported success.

We guess when node receive the udp packets from producer pod, route by stale conntrack entry as @kammous show.

**Comment by kammous** [comment]
> 
> 
> @lmm I am seeing this issue when I deployed Nginx Ingress controller with Mettallb hostnetwork. This is a major issue if there is only a single src / dst ip and src/dst port traffic with a daemonset.
> -> Same node chosen on pod restart
> -> Stale entry not allowing traffic to flow through
> -> Post timeout or force delete of conntrack entry its working fine..
> 
> @karmmous Did you find any solution for the same ?

@sunfuze , sorry for the late reponse. I have had to switch kube-proxy mode from iptables to ipvs to work around the issue.
If you have the same config, it is worth giving it a try. The same for you @workhorse2020.

**Comment by kammous** [comment]
> @kammous We are facing a similar issue with some differences. Indeed, changing kube-proxy to ipvs works as a WA but did you manage to raise the issue with kubernetes since this seems to be related with kube-proxy?

I'm suspecting the network plugin because I wasn't able to reproduce it with flannel.
The similar kube-proxy issue mentioned above is closed and no one is complaining since so I tend to think it is not the culprit.
